File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/model/WebViewModel.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.model;

import android.os.Bundle;

import com.urbanairship.android.layout.event.WebViewEvent;
import com.urbanairship.android.layout.property.Border;
import com.urbanairship.android.layout.property.Color;
import com.urbanairship.android.layout.property.ViewType;
import com.urbanairship.android.layout.reporting.LayoutData;
import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class WebViewModel extends BaseModel {
    @NonNull
    private final String url;

    @Nullable
    private Bundle savedState;

    public WebViewModel(@NonNull String url, @Nullable Color backgroundColor, @Nullable Border border) {
        super(ViewType.WEB_VIEW, backgroundColor, border);

        this.url = url;
    }

    @NonNull
    public static WebViewModel fromJson(@NonNull JsonMap json) throws JsonException {
        String url = json.opt("url").optString();
        Color backgroundColor = backgroundColorFromJson(json);
        Border border = borderFromJson(json);

        return new WebViewModel(url, backgroundColor, border);
    }

    @NonNull
    public String getUrl() {
        return url;
    }

    public void onClose() {
        bubbleEvent(new WebViewEvent.Close(), LayoutData.empty());
    }

    public void saveState(@NonNull Bundle bundle) {
        savedState = bundle;
    }

    @Nullable
    public Bundle getSavedState() {
        return savedState;
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/model/WebViewModel.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.model

import android.content.Context
import android.os.Bundle
import com.urbanairship.android.layout.environment.LayoutEvent
import com.urbanairship.android.layout.environment.ModelEnvironment
import com.urbanairship.android.layout.environment.ViewEnvironment
import com.urbanairship.android.layout.event.ReportingEvent
import com.urbanairship.android.layout.info.WebViewInfo
import com.urbanairship.android.layout.property.EventHandler
import com.urbanairship.android.layout.property.hasTapHandler
import com.urbanairship.android.layout.view.WebViewView
import kotlin.time.Duration.Companion.milliseconds
import kotlinx.coroutines.launch

internal class WebViewModel(
    viewInfo: WebViewInfo,
    environment: ModelEnvironment,
    properties: ModelProperties
) : BaseModel<WebViewView, WebViewInfo, BaseModel.Listener>(
    viewInfo = viewInfo,
    environment = environment,
    properties = properties
) {

    var savedState: Bundle? = null

    override fun onCreateView(
        context: Context,
        viewEnvironment: ViewEnvironment,
        itemProperties: ItemProperties?
    ) = WebViewView(context, this, viewEnvironment).apply {
        id = viewId
    }

    override fun onViewAttached(view: WebViewView) {
        if (viewInfo.eventHandlers.hasTapHandler()) {
            viewScope.launch {
                view.taps().collect { handleViewEvent(EventHandler.Type.TAP) }
            }
        }
    }

    fun onClose() {
        report(
            event = ReportingEvent.Dismiss(
                data = ReportingEvent.DismissData.UserDismissed,
                displayTime = environment.displayTimer.time.milliseconds,
                context = layoutState.reportingContext()
            )
        )
        broadcast(LayoutEvent.Finish())
    }
}
